src/compiler/scanner.ts

```diff
@@ -2384,15 +2444,979 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
+                    break;
+                }
+                const flag = characterToRegularExpressionFlag(String.fromCharCode(ch));
+                if (flag === undefined) {
+                    error(Diagnostics.Unknown_regular_expression_flag, p, 1);
+                }
+                else if (regExpFlags & flag) {
+                    error(Diagnostics.Duplicate_regular_expression_flag, p, 1);
+                }
+                else if (((regExpFlags | flag) & RegularExpressionFlags.UnicodeMode) === RegularExpressionFlags.UnicodeMode) {
+                    error(Diagnostics.The_Unicode_u_flag_and_the_Unicode_Sets_v_flag_cannot_be_set_simultaneously, p, 1);
+                }
+                else {
+                    regExpFlags |= flag;
+                    const availableFrom = regularExpressionFlagToFirstAvailableLanguageVersion(flag)!;
+                    if (languageVersion < availableFrom) {
+                        error(Diagnostics.This_regular_expression_flag_is_only_available_when_targeting_0_or_later, p, 1, ScriptTarget[availableFrom]);
+                    }
+                }
                 p++;
             }
-            pos = p;
+            pos = tokenStart + 1;
+            const saveTokenPos = tokenStart;
+            const saveTokenFlags = tokenFlags;
+            scanRegularExpressionWorker(text.slice(0, endOfBody), endOfBody, regExpFlags, isUnterminated);
+            if (!isUnterminated) {
+                pos = p;
+            }
+            tokenStart = saveTokenPos;
+            tokenFlags = saveTokenFlags;
             tokenValue = text.substring(tokenStart, pos);
             token = SyntaxKind.RegularExpressionLiteral;
         }
         return token;
+
+        function scanRegularExpressionWorker(text: string, end: number, regExpFlags: RegularExpressionFlags, isUnterminated: boolean) {
+            const unicodeMode = !!(regExpFlags & RegularExpressionFlags.UnicodeMode);
+            const unicodeSetsMode = !!(regExpFlags & RegularExpressionFlags.UnicodeSets);
+            let mayContainStrings = false;
+
+            let numberOfCapturingGroups = 0;
+            const groupSpecifiers = new Set<string>();
+            const groupNameReferences: (TextRange & { name: string; })[] = [];
+            const decimalEscapes: (TextRange & { value: number; })[] = [];
+            const namedCapturingGroups: Set<string>[] = [];
+
+            // Disjunction ::= Alternative ('|' Alternative)*
+            function scanDisjunction(isInGroup: boolean) {
+                while (true) {
+                    namedCapturingGroups.push(new Set());
+                    scanAlternative(isInGroup);
+                    namedCapturingGroups.pop();
+                    if (text.charCodeAt(pos) !== CharacterCodes.bar) {
+                        return;
+                    }
+                    pos++;
+                }
+            }
+
+            // Alternative ::= Term*

```

Comment:
I'd suggest that if we're going to include the grammar in a comment that we use the `grammarkdown`-format grammar from the spec itself as it makes it easier to compare against the spec.

Reply:
I actually did some sort of “elaboration” of the spec such as inlining intermediate productions instead of merely copying, aiming to make it more easily understandable at a glance
Both have their own advantages I believe